### PR TITLE
Add a very basic many to many pubsub

### PR DESCRIPTION
**Pull request by whyrusleeping**
This is experimental, but will very likely be merged into master.

For more information on the general protocol, take a look at: https://github.com/libp2p/pubsub-notes

and for the implementation being used here: https://github.com/whyrusleeping/go-floodsub

### CODE REVIEW

File: core/commands/pubsub.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+var PubsubPubCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Publish a message to a given pubsub sub topic.",
+		ShortDescription: `pubsub sub pub pub hubbub sub?
+pub. Pubsub subsub subpubsub hubbub pub sub grub subhubbub.
+Pubsubs sub hubbub sub tub, dub dub dub wub wub wub wub wub. Pubsub.
+Pubsub subsub pubsub hubbub dubtub? Wub. Wub. Pubsub.
+Nubs, Tubs, Wubs, pub Dubs. Pub subdub pub pub sub nub.
+
+    ipfs pubsub pub "pubsub hubbub"
```
Comment (whyrusleeping) [inline core/commands/pubsub.go]: whoops, forgot the topic flag here

**Comment by whyrusleeping** [comment]
Voting for the name, thumbs up comments to vote:

**Comment by whyrusleeping** [comment]
OPTION: floodsub

**Comment by whyrusleeping** [comment]
OPTION: hubbub

**Comment by whyrusleeping** [comment]
OPTION: libp2p-flooding-based-publish-subscribe-system

**Comment by haadcode** [comment]
OPTION: hub

File: core/commands/pubsub.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	Type: floodsub.Message{},
+}
+var PubsubPubCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Publish a message to a given pubsub sub topic.",
+		ShortDescription: `pubsub sub pub pub hubbub sub?
+pub. Pubsub subsub subpubsub hubbub pub sub grub subhubbub.
+Pubsubs sub hubbub sub tub, dub dub dub wub wub wub wub wub. Pubsub.
+Pubsub subsub pubsub hubbub dubtub? Wub. Wub. Pubsub.
+Nubs, Tubs, Wubs, pub Dubs. Pub subdub pub pub sub nub.
```
Comment (jbenet) [inline core/commands/pubsub.go]: https://www.youtube.com/watch?v=TlYaupCTzBY

File: core/commands/pubsub.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+	cmds "github.com/ipfs/go-ipfs/commands"
+
+	floodsub "gx/ipfs/QmQriRMW5cCJyLrzDnXi7fZ5mVbetiEZjPjbqoJhuSL94m/floodsub"
+	u "gx/ipfs/QmZNVWh8LLjAavuQ2JXuFmuYH3C11xo988vSgp7UQrTRj1/go-ipfs-util"
+)
+
+var PubsubCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Only pubsub by the vaguest technical definition",
```
Comment (jbenet) [inline core/commands/pubsub.go]: Add a description, denoting experimental

File: core/commands/pubsub.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	},
+	Subcommands: map[string]*cmds.Command{
+		"pub": PubsubPubCmd,
+		"sub": PubsubSubCmd,
+	},
+}
+var PubsubSubCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "subscribe to messages on a given topic",
```
Comment (jbenet) [inline core/commands/pubsub.go]: Add a description, denoting experimental

File: core/commands/pubsub.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("topic", true, false, "String name of topic to subscribe to."),
```
Comment (jbenet) [inline core/commands/pubsub.go]: options for other formats? as discussed in https://github.com/libp2p/pubsub-notes/blob/master/flooding/cli-ux

File: core/commands/pubsub.go (hunk 2 above)
Comment (jbenet) [inline core/commands/pubsub.go]: fix desc. Add a description, denoting experimental

File: core/commands/pubsub.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    ipfs pubsub pub "pubsub hubbub"
+
+Pub: hubbub pub sub sub tub pub bub.
+		`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("data", true, false, "Data to send to david dias. (and only him)").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.StringOption("topic", "t", "Topic to pubusb to."),
```
Comment (jbenet) [inline core/commands/pubsub.go]: this should be an Argument

File: core/commands/pubsub.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+Pubsub subsub pubsub hubbub dubtub? Wub. Wub. Pubsub.
+Nubs, Tubs, Wubs, pub Dubs. Pub subdub pub pub sub nub.
+
+    ipfs pubsub pub "pubsub hubbub"
+
+Pub: hubbub pub sub sub tub pub bub.
+		`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("data", true, false, "Data to send to david dias. (and only him)").EnableStdin(),
```
Comment (jbenet) [inline core/commands/pubsub.go]: input opts, as in https://github.com/libp2p/pubsub-notes/blob/master/flooding/cli-ux

**Comment by whyrusleeping** [comment]
@jbenet should we have the command set to be enabled by a config option? Or do you think that just stating in each commands helptext that its experimental is enough?

**Comment by jbenet** [comment]
> should we have the command set to be enabled by a config option? Or do you think that just stating in each commands helptext that its experimental is enough?

Hmmm. yeah i'd hide them unless either:
-  a flag like `ipfs daemon --experiment-pubsub` or something, that flips a switch and everything is there.
- a config option like `Experimental.PubSub = true`

**Comment by lgierth** [comment]
> -  a flag like `ipfs daemon --experiment-pubsub` or something, that flips a switch and everything is there.
> - a config option like `Experimental.PubSub = true`

a flag would be better because a config option is hard to remove once it's there.

**Comment by djdv** [comment]
What about "subhub"?

**Comment by jbenet** [comment]
2016-09-12 at 6 08 pm

**Comment by jbenet** [comment]
# it works :neckbeard: 👍

**Comment by whyrusleeping** [comment]
Rebased on top of new libp2p changes that let us interop with js-ipfs. Also cleaned up a bunch of the code

**Comment by whyrusleeping** [comment]
Alright, merging this in. Its experimental, definitely subject to change, but man is it fun to play with.

**Comment by elimisteve** [comment]
@whyrusleeping Sweet! Are there any Go code examples using this functionality that I can check out?

**Comment by whyrusleeping** [comment]
Not yet! It's on my to-do list :)

**Comment by leeola** [comment]
Also some notes on performance. I just stumbled upon this _(had no idea!)_, and am really interested. With that said, i've got no idea how well it scales, very broad performance characteristics, etc.

A high level paper would be very appreciated! Excellent work 💯 !

**Comment by kikoncuo** [comment]
@whyrusleeping That is amazing, are you still planning to do any examples?

**Pull request merged**